parastell/magnet_coils.py

```diff
@@ -401,164 +319,104 @@ class MagnetCoil(object):
-
-        Arguments:
-            path_origin (int): index of initial point in filament path.
-            surf_id (int): index of cross-section surface.
-            t_vec (list of float): cross-section thickness vector.
-            norm (list of float): cross-section normal vector.
-            rot_axis (list of float): axis about which to rotate the
-                cross-section.
-            rot_ang_norm (float): angle by which cross-section was rotated to
-                align its normal with the initial point tangent (deg).
-        """
-        # Determine orientation of thickness vector after cross-section was
-        # oriented along filament origin tangent
+    def __init__(self, coords, center_of_mass, width, thickness, sample_mod):
 
-        # Compute part of thickness vector parallel to rotation axis
-        t_vec_par = normalize(np.inner(t_vec, rot_axis) * rot_axis)
-        # Compute part of thickness vector orthogonal to rotation axis
-        t_vec_perp = normalize(t_vec - t_vec_par)
+        self.sample_mod = sample_mod
+        self.coords = coords
+        self.center_of_mass = center_of_mass
+        self.width = width
+        self.thickness = thickness
 
-        # Compute vector othogonal to both rotation axis and orthogonal
-        # part of thickness vector
-        orth = normalize(np.cross(rot_axis, t_vec_perp))
-
-        # Determine part of rotated vector parallel to original
-        rot_par = np.cos(rot_ang_norm)
-        # Determine part of rotated vector orthogonal to original
-        rot_perp = np.sin(rot_ang_norm)
-
-        # Compute orthogonal part of thickness vector after rotation
-        t_vec_perp_rot = rot_par * t_vec_perp + rot_perp * orth
-        # Compute thickness vector after rotation
-        t_vec_rot = normalize(t_vec_perp_rot + t_vec_par)
-
-        # Orient cross-section in its plane such that it faces the global origin
-
-        # Extract initial path point
-        pos = cubit.vertex(path_origin).coordinates()
-
-        # Project position vector onto cross-section
-        pos_proj = normalize(pos - np.inner(pos, norm) * norm)
-
-        # Compute angle by which to rotate cross-section such that it faces the
-        # origin
-        rot_ang_orig = np.arccos(np.inner(pos_proj, t_vec_rot))
-
-        # Re-orient rotated cross-section such that thickness vector faces
-        # origin
-        cubit.cmd(
-            f"rotate Surface {surf_id} angle {np.rad2deg(rot_ang_orig)} about "
-            "origin 0 0 0 direction " + " ".join(str(i) for i in norm)
+    @property
+    def coords(self):
+        return self._coords
+
+    @coords.setter
+    def coords(self, data):
+        # Compute tangents
+        tangents = np.subtract(
+            np.append(data[1:], [data[1]], axis=0),
+            np.append([data[-2]], data[0:-1], axis=0),
         )
+        tangents = tangents / np.linalg.norm(tangents, axis=1)[:, np.newaxis]
```

Comment:
I think there may be some confusion here. What I wrote computes, at each filament point, the normalized direction from the previous point to the next point. Computation of the filament normals and binormals relies on the tangents being unit vectors. From what I infer, what has been suggested attempts to compute essentially the same thing, except that the tangents are not normalized at the end. I know I mentioned previously that the "central difference approximation" is used to compute tangents, but that term was used loosely to convey which points are used to compute the tangent rather than an actual derivative.

It may also help to remember that for each filament, the initial and final points are the same. So, when computing the tangent for each of those points, we want to compute the difference between the second and second-to-last points in the array. That is why we add the second point to the end of the first array and second-to-last point to the beginning of the second array in the `np.subtract` operation. We are essentially just rearranging the loops, shifting the first and second arrays one index forward and one index backward, respectively.

Adding the first point to the end of the first array and last point to the beginning of the second array, as suggested, will lead to zero vectors computed for the tangents at both the first and final points.

Reply:
I see where you're coming from now, I wasn't even thinking of central difference as an average of the forward and backward differences but rather just as the "end product" where the point of interest cancels out; apologies for that. I think the implementation I wrote is straight-forward and does what we want.